File: pcdaemon/security/message_security_preprocessor.py
from security.guard import Guard
from security.session import Session
from security.tls_packet import TLSPacket
from utils.msg_codes import TLSCode


class MessageSecurityPreprocessor:

    def __init__(self, guard: Guard):
        self.guard = guard

    def preprocess_to_send(self, session: Session, message_to_send: bytes) -> bytes:
        nonce = self.guard.get_nonce()

        tls_packet = TLSPacket(TLSCode.SECURE, len(
            message_to_send), len(nonce), nonce, message_to_send)

        encrypted_data = self.guard.encrypt(
            session.get_tcp_secret_key(), tls_packet.data, tls_packet.header, nonce)

        return TLSPacket.build_from_header_and_data(tls_packet.header, encrypted_data).full

    def preprocess_received(self, session: Session, received_message: bytes) -> bytes:
        tls_packet = TLSPacket.build_from_raw(received_message)

        assert len(tls_packet.nonce) == self.guard.get_nonce_length()

        return self.guard.decrypt(session.get_tcp_secret_key(), tls_packet.data, tls_packet.header, tls_packet.nonce)

    def get_basic_header_size(self) -> int:
        return TLSPacket.HEADER_SIZE

    def get_message_size(self, basic_header: bytes) -> int:
        return TLSPacket.get_packet_size(basic_header, self.guard.get_tag_length())

File: app/src/main/java/com/example/pilot/security/MessageSecurityPreprocessor.java
package com.example.pilot.security;

import android.util.Log;

import com.example.pilot.security.exceptions.AuthenticationException;
import com.example.pilot.security.exceptions.SecurityException;
import com.example.pilot.security.utils.TLSCode;
import com.example.pilot.security.utils.TLSPacket;

import java.nio.charset.StandardCharsets;
import java.security.SecureRandom;

public class MessageSecurityPreprocessor {
    private final static String TAG = "Message Security Preprocessor";
    private final Guard guard;

    public MessageSecurityPreprocessor(Guard guard) {
        this.guard = guard;
    }

    public byte[] preprocessToSend(byte[] messageToSend) throws SecurityException {
        byte[] nonce = guard.getNonce();
        TLSPacket tlsPacket = new TLSPacket(TLSCode.SECURE, (short) messageToSend.length, nonce.length, nonce, messageToSend);

        Log.d(TAG, "sending len: " + messageToSend.length);

        byte[] encryptedData = guard.encrypt(tlsPacket.data, tlsPacket.header, nonce);

        return new TLSPacket(tlsPacket.header, encryptedData).full;
    }

    // returns decrypted data from underlying protocol, with all security-layer data stripped
    public byte[] preprocessReceived(byte[] receivedMessage) throws SecurityException, AuthenticationException {
        TLSPacket tlsPacket = new TLSPacket(receivedMessage);

        if (tlsPacket.code != TLSCode.SECURE)
            throw new SecurityException("Expected encrypted packet but got code " + tlsPacket.code);

        byte[] nonce = tlsPacket.nonce;

        if (nonce.length != guard.getNonceLength())
            throw new SecurityException("Invalid nonce length, expected " + guard.getNonceLength() +
                    " got " + nonce.length + "\nFrom: " + new String(tlsPacket.full, StandardCharsets.UTF_8));

        return guard.decrypt(tlsPacket.data, tlsPacket.header, nonce);
    }


    public int getBasicHeaderSize() {
        return TLSPacket.HEADER_SIZE;
    }

    public int getMessageSize(byte[] basicHeader) {
        // first HEADER_SIZE bytes
        return TLSPacket.getMessageSize(basicHeader, guard.getTagLength());
    }

    public void setKey(byte[] decoded) {
        guard.setSessionKey(decoded);
    }
}
